### ISSUE TITLE
`update_cells` corrupts data if input is unsorted

### ISSUE THREAD

**Issue by sbrudenell** [burnash/gspread#702]
Found in version 3.1.0.

`gspread.utils.cell_list_to_rect` assumes that the first element of the input list should be used as the "offset" for row and col. If it's not, it constructs a garbage rect using some unintended negative indexes.

This results in `update_cells` corrupting the sheet.

Example:

```python
In [1]: from gspread import utils, models

In [2]: utils.cell_list_to_rect([models.Cell(1, 1, "a1"), models.Cell(2, 2, "b2")])
Out[2]: [['a1', None], [None, 'b2']]

In [3]: utils.cell_list_to_rect([models.Cell(2, 2, "b2"), models.Cell(1, 1, "a1")])
Out[3]: [['b2']]
```

**Comment by sbrudenell** [comment]
It gets worse: if `update_cells` is called without updating a top-left corner of a rect, there's no way to construct the input sanely.

```python
In [1]: from gspread import utils, models

In [2]: utils.cell_list_to_rect([models.Cell(1, 2, "a2"), models.Cell(2, 1, "b1")])
Out[2]: [['a2'], [None]]

In [3]: utils.cell_list_to_rect([models.Cell(2, 1, "b1"), models.Cell(1, 2, "a2")])
Out[3]: [['b1', None]]
```

**Comment by MarsCapone** [comment]
I've run into this also, but I need to resolve it for a client. Ideally I don't want to change library.

As a further demonstration of the issue...

```
In [1]: cells
Out[1]:
[<Cell R1C1 0>,
 <Cell R1C2 1>,
 <Cell R1C3 2>,
 <Cell R2C1 3>,
 <Cell R2C2 4>,
 <Cell R2C3 5>,
 <Cell R3C1 6>,
 <Cell R3C2 7>,
 <Cell R3C3 8>]

In [2]: utils.cell_list_to_rect(cells)
Out[2]: [[0, 1, 2], [3, 4, 5], [6, 7, 8]]
```

That's what we want to happen if those cells are already in my sheet. 

Now, we'll swap the first two cells. They're just in a different order in the list, but now the first column is missing.
```
In [3]: cells2 = cells[:]

In [4]: cells2[0], cells2[1] = cells2[1], cells2[0]

In [5]: utils.cell_list_to_rect(cells2)
Out[5]: [[1, 2], [4, 5], [7, 8]]
```

I though we could just sort the list by row then column...
```
In [6]: utils.cell_list_to_rect(sorted(cells2, key=lambda x: (x.row, x.col)))
Out[6]: [[0, 1, 2], [3, 4, 5], [6, 7, 8]]
```

But that doesn't work is there is no top left corner
```
In [7]: cells3 = cells[1:]

In [8]: utils.cell_list_to_rect(cells3)
Out[8]: [[1, 2], [4, 5], [7, 8]]

In [9]: utils.cell_list_to_rect(sorted(cells3, key=lambda x: (x.row, x.col)))
Out[9]: [[1, 2], [4, 5], [7, 8]]

In [10]: utils.cell_list_to_rect(sorted(cells3, key=lambda x: (x.row, x.col)))
Out[10]: [[1, 2], [4, 5]]
```

---

It seems like the only solution is to calculate the result correctly. The only way I can see to do this is to loop through once to find the extents. Then loop through another time to build the correct array. I don't think this is so different from the current implementation.

**Comment by taewookim** [comment]
anyone with update on this issue?

**Comment by bool-** [comment]
> 
> 
> anyone with update on this issue?

Simply calling `list.sort()` or `sorted()` should resolve it assuming you have a top left corner, otherwise the function would need to be modified to calculate the top-left corner of the bounding box and add it to the `cell_list` prior to the main work of the `cell_list_to_rect` function.

@MarsCapone it should be simpler to just add the top left corner by assumption, example below.
```
from operator import attrgetter
min_row = min(cell_list, key=attrgetter('row'))
min_col = min(cell_list, key=attrgetter('col'))
cell_list.
```

**Comment by burnash** [comment]
Thanks for reporting this. The update from @bool- is correct, this is fixed in #613 and released in version 3.2.0.
